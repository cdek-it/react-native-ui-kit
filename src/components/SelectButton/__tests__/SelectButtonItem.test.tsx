import { IconArrowDownRight } from '@tabler/icons-react-native'
import { act, render, userEvent, waitFor } from '@testing-library/react-native'

import { Pressable } from 'react-native'
import { useSharedValue } from 'react-native-reanimated'

import {
  type SelectButtonItemProps,
  SelectButtonItem,
} from '../SelectButtonItem'

type SelectButtonItemTestProps = Partial<
  Omit<SelectButtonItemProps, 'position'>
> & { readonly position?: number; readonly withButton?: boolean }

const TestComponent = ({
  position: positionProp = 0,
  onPress = jest.fn(),
  index = 0,
  withButton,
  ...rest
}: SelectButtonItemTestProps) => {
  const position = useSharedValue(positionProp)

  return (
    <>
      <SelectButtonItem
        index={index}
        position={position}
        onPress={onPress}
        {...rest}
      />
      {withButton ? (
        <Pressable
          testID='ChangePosition'
          onPress={() => {
            position.value += 1
          }}
        />
      ) : null}
    </>
  )
}

describe('SelectButtonItem', () => {
  beforeEach(() => {
    jest.useFakeTimers()
  })

  afterEach(() => {
    jest.runOnlyPendingTimers()
    jest.useRealTimers()
  })

  const snapshotCases: Array<[SelectButtonItemTestProps]> = [
    [{ label: 'ButtonSelect' }],
    [{ size: 'small', Icon: IconArrowDownRight }],
    [{ size: 'base', Icon: IconArrowDownRight, showIcon: false }],
    [
      {
        size: 'large',
        Icon: IconArrowDownRight,
        showIcon: true,
        label: 'ButtonSelect',
        disabled: true,
      },
    ],
    [
      {
        size: 'xlarge',
        label: 'ButtonSelect',
        Icon: IconArrowDownRight,
        index: 1,
      },
    ],
  ]

  test.each(snapshotCases)('%p', async (props) => {
    const { toJSON } = render(<TestComponent {...props} />)
    await act(async () => {
      jest.advanceTimersByTime(600)

      expect(toJSON()).toMatchSnapshot()
    })
  })

  test('handle press', async () => {
    const mockedOnPress = jest.fn()
    const { queryByTestId } = render(<TestComponent onPress={mockedOnPress} />)
    const user = userEvent.setup()

    const touchableOpacity = queryByTestId('SelectButtonItem_TouchableOpacity')
    await user.press(touchableOpacity)

    expect(mockedOnPress).toHaveBeenCalled()
  })

  test('position change', async () => {
    const mockedOnPress = jest.fn()
    const { queryAllByTestId } = render(
      <TestComponent
        withButton
        Icon={IconArrowDownRight}
        onPress={mockedOnPress}
      />
    )
    const user = userEvent.setup()
    const pressable = queryAllByTestId('ChangePosition')
    let icons = queryAllByTestId('SelectButtonItem_Icon')

    await waitFor(() => expect(icons[0]).toHaveStyle({ color: '#188700' }))

    await user.press(pressable[0])
    jest.advanceTimersByTime(600)

    icons = queryAllByTestId('SelectButtonItem_Icon')

    expect(icons[0]).toHaveStyle({ color: 'rgba(0, 0, 0, 0.6000)' })
  })
})
