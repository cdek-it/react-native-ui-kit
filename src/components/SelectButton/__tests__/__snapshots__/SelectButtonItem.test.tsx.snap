// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectButtonItem {"Icon": [Object], "disabled": true, "label": "ButtonSelect", "showIcon": true, "size": "large"} 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "borderColor": "#dbdada",
      "borderRadius": 3.5,
      "borderWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "gap": 10.5,
      "height": 49,
      "justifyContent": "center",
      "opacity": 1,
    }
  }
  testID="SelectButtonItem_TouchableOpacity"
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={21}
    bbWidth={21}
    fill="none"
    focusable={false}
    height={21}
    meetOrSlice={0}
    minX={0}
    minY={0}
    stroke="currentColor"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={2}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        [
          {
            "color": "rgba(0, 0, 0, 0.6000)",
          },
          false,
          {
            "color": "#dbdada",
          },
        ],
        {
          "flex": 0,
          "height": 21,
          "width": 21,
        },
      ]
    }
    testID="SelectButtonItem_Icon"
    vbHeight={24}
    vbWidth={24}
    width={21}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      color="#dbdada"
      fill={null}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeLinecap",
          "strokeLinejoin",
        ]
      }
      stroke={
        {
          "type": 2,
        }
      }
      strokeLinecap={1}
      strokeLinejoin={1}
      strokeWidth={2}
    >
      <RNSVGPath
        color="#dbdada"
        d="M7 7l10 10"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
        testID="SelectButtonItem_Icon"
      />
      <RNSVGPath
        color="#dbdada"
        d="M17 8l0 9l-9 0"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
        testID="SelectButtonItem_Icon"
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <Text
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    numberOfLines={1}
    style={
      {
        "color": "#dbdada",
        "flexShrink": 1,
        "fontSize": 17.5,
        "fontWeight": 700,
      }
    }
    testID="SelectButtonItem_Text"
  >
    ButtonSelect
  </Text>
</View>
`;

exports[`SelectButtonItem {"Icon": [Object], "index": 1, "label": "ButtonSelect", "size": "xlarge"} 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "borderColor": "transparent",
      "borderWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "gap": 10.5,
      "height": 56,
      "justifyContent": "center",
      "opacity": 1,
    }
  }
  testID="SelectButtonItem_TouchableOpacity"
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={28}
    bbWidth={28}
    fill="none"
    focusable={false}
    height={28}
    meetOrSlice={0}
    minX={0}
    minY={0}
    stroke="currentColor"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={2}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        [
          {
            "color": "rgba(0, 0, 0, 0.6000)",
          },
          false,
          undefined,
        ],
        {
          "flex": 0,
          "height": 28,
          "width": 28,
        },
      ]
    }
    testID="SelectButtonItem_Icon"
    vbHeight={24}
    vbWidth={24}
    width={28}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      color="rgba(0, 0, 0, 0.6000)"
      fill={null}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeLinecap",
          "strokeLinejoin",
        ]
      }
      stroke={
        {
          "type": 2,
        }
      }
      strokeLinecap={1}
      strokeLinejoin={1}
      strokeWidth={2}
    >
      <RNSVGPath
        color="rgba(0, 0, 0, 0.6000)"
        d="M7 7l10 10"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
        testID="SelectButtonItem_Icon"
      />
      <RNSVGPath
        color="rgba(0, 0, 0, 0.6000)"
        d="M17 8l0 9l-9 0"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
        testID="SelectButtonItem_Icon"
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <Text
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "color": "rgba(0, 0, 0, 0.6)",
        },
      }
    }
    numberOfLines={1}
    style={
      {
        "color": "rgba(0, 0, 0, 0.6)",
        "flexShrink": 1,
        "fontSize": 21,
        "fontWeight": 700,
      }
    }
    testID="SelectButtonItem_Text"
  >
    ButtonSelect
  </Text>
</View>
`;

exports[`SelectButtonItem {"Icon": [Object], "showIcon": false, "size": "base"} 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "borderColor": "transparent",
      "borderWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "gap": 7,
      "height": 35,
      "justifyContent": "center",
      "opacity": 1,
    }
  }
  testID="SelectButtonItem_TouchableOpacity"
>
  <Text
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "color": "rgba(24, 135, 0, 1)",
        },
      }
    }
    numberOfLines={1}
    style={
      {
        "color": "rgba(24, 135, 0, 1)",
        "flexShrink": 1,
        "fontSize": 14,
        "fontWeight": 700,
      }
    }
    testID="SelectButtonItem_Text"
  />
</View>
`;

exports[`SelectButtonItem {"Icon": [Object], "size": "small"} 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "borderColor": "transparent",
      "borderWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "gap": 7,
      "height": 28,
      "justifyContent": "center",
      "opacity": 1,
    }
  }
  testID="SelectButtonItem_TouchableOpacity"
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight={14}
    bbWidth={14}
    fill="none"
    focusable={false}
    height={14}
    meetOrSlice={0}
    minX={0}
    minY={0}
    stroke="currentColor"
    strokeLinecap="round"
    strokeLinejoin="round"
    strokeWidth={2}
    style={
      [
        {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        [
          {
            "color": "rgba(0, 0, 0, 0.6000)",
          },
          false,
          undefined,
        ],
        {
          "flex": 0,
          "height": 14,
          "width": 14,
        },
      ]
    }
    testID="SelectButtonItem_Icon"
    vbHeight={24}
    vbWidth={24}
    width={14}
    xmlns="http://www.w3.org/2000/svg"
  >
    <RNSVGGroup
      color="rgba(0, 0, 0, 0.6000)"
      fill={null}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeLinecap",
          "strokeLinejoin",
        ]
      }
      stroke={
        {
          "type": 2,
        }
      }
      strokeLinecap={1}
      strokeLinejoin={1}
      strokeWidth={2}
    >
      <RNSVGPath
        color="rgba(0, 0, 0, 0.6000)"
        d="M7 7l10 10"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
        testID="SelectButtonItem_Icon"
      />
      <RNSVGPath
        color="rgba(0, 0, 0, 0.6000)"
        d="M17 8l0 9l-9 0"
        fill={null}
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
            "strokeLinecap",
            "strokeLinejoin",
          ]
        }
        stroke={
          {
            "type": 2,
          }
        }
        strokeLinecap={1}
        strokeLinejoin={1}
        strokeWidth={2}
        testID="SelectButtonItem_Icon"
      />
    </RNSVGGroup>
  </RNSVGSvgView>
  <Text
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "color": "rgba(24, 135, 0, 1)",
        },
      }
    }
    numberOfLines={1}
    style={
      {
        "color": "rgba(24, 135, 0, 1)",
        "flexShrink": 1,
        "fontSize": 12.25,
        "fontWeight": 700,
      }
    }
    testID="SelectButtonItem_Text"
  />
</View>
`;

exports[`SelectButtonItem {"label": "ButtonSelect"} 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "borderColor": "transparent",
      "borderWidth": 1,
      "flex": 1,
      "flexDirection": "row",
      "gap": 7,
      "height": 35,
      "justifyContent": "center",
      "opacity": 1,
    }
  }
  testID="SelectButtonItem_TouchableOpacity"
>
  <Text
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "color": "rgba(24, 135, 0, 1)",
        },
      }
    }
    numberOfLines={1}
    style={
      {
        "color": "rgba(24, 135, 0, 1)",
        "flexShrink": 1,
        "fontSize": 14,
        "fontWeight": 700,
      }
    }
    testID="SelectButtonItem_Text"
  >
    ButtonSelect
  </Text>
</View>
`;
